File: lib/parsers/python-parser.js
const { execSync } = require('child_process');
const path = require('path');
const fs = require('fs');
const os = require('os');

class PythonParser {
  /**
   * Synchronous version of parse for compatibility with existing codebase
   * @param {string} content - Python source code
   * @param {string} filePath - Path to the file
   * @returns {Array} Array of symbol names (for compatibility)
   */
  parseSync(content, filePath) {
    try {
      // Create temporary files
      const tempDir = os.tmpdir();
      const timestamp = Date.now() + Math.random().toString(36).substr(2, 9);
      const sourceFile = path.join(tempDir, `python_source_${timestamp}.py`);
      const scriptFile = path.join(tempDir, `python_parser_${timestamp}.py`);
      
      // Write the source code to a temporary file
      fs.writeFileSync(sourceFile, content, 'utf-8');
      
      // Create the parser script with escaped path
      const escapedSourcePath = sourceFile.replace(/\\/g, '\\\\').replace(/'/g, "\\'");
      const parserScript = `import ast
import json
import sys

def extract_symbols(file_path):
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            source_code = f.read()
        
        tree = ast.parse(source_code)
        symbols = []
        
        # Only process top-level nodes, not nested ones
        for node in tree.body:
            if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
                if not node.name.startswith('_'):
                    symbols.append(node.name)
            elif isinstance(node, ast.ClassDef):
                if not node.name.startswith('_'):
                    symbols.append(node.name)
            elif isinstance(node, ast.Assign) and hasattr(node, 'lineno'):
                for target in node.targets:
                    if isinstance(target, ast.Name):
                        if not target.id.startswith('_'):
                            symbols.append(target.id)
        
        return symbols
    except Exception as e:
        return []

try:
    symbols = extract_symbols('${escapedSourcePath}')
    print(json.dumps(symbols))
except Exception as e:
    print(json.dumps([]))
`;

      fs.writeFileSync(scriptFile, parserScript, 'utf-8');

      const output = execSync(`python3 "${scriptFile}"`, {
        encoding: 'utf-8',
        timeout: 5000,
        stdio: ['pipe', 'pipe', 'pipe']
      });

      // Clean up temporary files
      try {
        fs.unlinkSync(sourceFile);
        fs.unlinkSync(scriptFile);
      } catch (cleanupError) {
        // Ignore cleanup errors
      }

      const symbols = JSON.parse(output.trim());
      return symbols;
      
    } catch (error) {
      console.warn(`Python parsing failed for ${filePath}: ${error.message}`);
      return [];
    }
  }
}

module.exports = PythonParser;

File: lib/parsers/temp/python_parser.py

import ast
import sys
import json

def extract_symbols(source_code):
    """Extract functions, classes, and module-level variables from Python code"""
    try:
        tree = ast.parse(source_code)
        symbols = []
        
        for node in ast.walk(tree):
            # Extract function definitions
            if isinstance(node, ast.FunctionDef):
                # Skip private functions (starting with _)
                if not node.name.startswith('_'):
                    symbols.append({
                        'name': node.name,
                        'type': 'function',
                        'line': node.lineno
                    })
            
            # Extract class definitions
            elif isinstance(node, ast.ClassDef):
                # Skip private classes (starting with _)
                if not node.name.startswith('_'):
                    symbols.append({
                        'name': node.name,
                        'type': 'class',
                        'line': node.lineno
                    })
            
            # Extract module-level variable assignments
            elif isinstance(node, ast.Assign) and hasattr(node, 'lineno'):
                # Only get top-level assignments (not inside functions/classes)
                if node.col_offset == 0:
                    for target in node.targets:
                        if isinstance(target, ast.Name):
                            # Skip private variables (starting with _)
                            if not target.id.startswith('_'):
                                symbols.append({
                                    'name': target.id,
                                    'type': 'variable',
                                    'line': node.lineno
                                })
        
        return symbols
    
    except SyntaxError as e:
        return []
    except Exception as e:
        return []

if __name__ == '__main__':
    source_code = sys.stdin.read()
    symbols = extract_symbols(source_code)
    print(json.dumps(symbols))
